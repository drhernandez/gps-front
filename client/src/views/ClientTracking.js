/*global google*/
import React from "react";
import {
  Container,
  Row,
  Col,
  Card,
  CardHeader,
  CardBody,
  Form,
  FormSelect,
  Alert
} from "shards-react";
import NavbarSearch from "../components/common/search/NavbarSearch";
import PageTitle from "../components/common/PageTitle";
import Gmap from "./../components/common/maps/Gmap";
import { Marker } from "./../components/common/maps/Marker";
import Constants from "../utils/Constants";
import { UsersService, VehiclesService } from "../api/services";
import { to } from "await-to-js";
import "../styles/clientTracking.css";

const errorMessages = {
  generic: {
    type: Constants.Themes.ERROR,
    message: "Algo salió mal al intentar obtener los datos del usuario. Inténtelo de nuevo."
  },
  userNotFound: {
    type: Constants.Themes.WARNING,
    message: "Usuario no encontrado."
  }
}

class ClientTracking extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchBarDisabled: false,
      user: null,
      trackings: [],
      markers: [],
      center: null,
      zoom: null,
      alert: {
        visible: false,
        ...errorMessages.generic
      }
    }
    this.search = this.search.bind(this);
    this.handleDeviceOnChange = this.handleDeviceOnChange.bind(this);
  };

  componentWillUnmount() {
    clearInterval(this.dataPolling);
  }

  async search(e) {
    e.preventDefault();
    e.persist();

    const email = e.target["search-input"].value;
    this.setState({
      searchBarDisabled: true,
      user: null
    })

    const [err, user] = await to(UsersService.getUserByEmail(email));
    if (err != null) {
      this.setState({
        searchBarDisabled: false,
        alert: {
          visible: true,
          ...errorMessages.userNotFound
        },
        showSppiner: false
      })
      setTimeout(() => {
        this.setState({
          alert: {
            visible: false
          }
        });
      }, 4000);
    }
    else {
      let [err, results] = await to(VehiclesService.searchVehicles(user.id));
      if (err != null) {
        console.log(err);
        this.setState({
          searchBarDisabled: false,
          alert: {
            visible: true,
            ...errorMessages.generic
          }
        })
        setTimeout(() => {
          this.setState({
            alert: {
              visible: false
            }
          });
        }, 4000);
      }
      else {
        user.vehicles = results.data;
        this.setState({
          searchBarDisabled: false,
          user: user
        });
      }
    }
  }

  handleDeviceOnChange(event) {
    event.persist();
    this.getCurrentLocation(event.target.value);
    this.dataPolling = setInterval(
      () => {
        try {
          this.getCurrentLocation(event.target.value);
        } catch (error) {
        }
      }, 4000);
  }

  async getCurrentLocation(vehicleID) {
    const [err, location] = await to(VehiclesService.getCurrentLocation(vehicleID));
    if (!err && location) {
      const marker = new Marker(location.id, location.label, undefined, google.maps.Animation.DROP, location.lat, location.lng);
      this.setState({
        trackings: [location],
        markers: [marker],
        center: marker.position,
        zoom: 15,
        alert: {
          visible: false,
          body: {}
        }
      })
    }
  }

  render() {
    return (
      <Container fluid className="main-content-container px-4">
        {/* Page Header */}
        <Row noGutters className="page-header py-4">
          <PageTitle title="Tracking" subtitle="Clientes" className="text-sm-left mb-3 col-sm-12" />
        </Row>

        {
          this.state.alert.visible &&
          <Row noGutters className="w-100">
            <Alert className="my-auto w-100" open={this.state.alert.visible} theme={this.state.alert.type}>
              {this.state.alert.message}
            </Alert>
          </Row>
        }

        <Row noGutters className="py-4">
          <Card small className="mx-auto w-100 mb-4">
            <CardHeader className="mt-3">
              <Row>
                <Col>
                  <NavbarSearch disabled={this.state.searchBarDisabled} placeholder="Buscar clientes por email..." searchFunction={this.search} />
                </Col>
                <Col>
                  <Form className="add-new-post">
                    <FormSelect id="feInputState" defaultValue="default" disabled={this.state.user == null} onChange={this.handleDeviceOnChange}>
                      <option value="default" disabled>Elija un vehículo...</option>
                      {this.state.user && this.state.user.vehicles.map((vehicle, idx) => (
                        <option key={vehicle.id} value={vehicle.id}>{`${vehicle.brand} ${vehicle.brand_line} - ${vehicle.plate}`}</option>
                      ))}
                    </FormSelect>
                  </Form>
                </Col>
              </Row>
            </CardHeader>
            <CardBody>
              <Gmap
                isMarkerShown
                containerElement={<div style={{ height: '450px' }} />}
                mapElement={<div style={{ height: `100%` }} />}
                zoom={this.state.zoom}
                markers={this.state.markers}
                center={this.state.center}
              />
            </CardBody>
          </Card>
        </Row>
      </Container>
    );
  }
}

export default ClientTracking;