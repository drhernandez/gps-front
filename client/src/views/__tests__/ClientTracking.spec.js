import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

jest.mock("../../api/services/usersService", () => ({ getUserByEmail: jest.fn() }));
import UsersService from "../../api/services/usersService";
jest.mock("../../api/services/vehiclesService", () => ({ searchVehicles: jest.fn(), getCurrentLocation: jest.fn() }));
import VehiclesService from '../../api/services/vehiclesService';

import ClientTracking from '../ClientTracking';

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}

let userMock;
let vehiclesMock;
let locationMock;

describe('ClientTracking component', () => {
  beforeEach(async () => {
    jest.resetModules();

    userMock = {
      id: 1
    }

    vehiclesMock = {
      data: [
        {
          id: 1,
          brand: "Frod",
          brand_line: "Fiesta",
          plate: "AA111AA"
        }
      ],
      paging: {
        total: 1
      }
    }

    locationMock = {
      "id": 1,
      "device_id": 1,
      "lat": -31.422108,
      "lng": -64.18643
    }

    await UsersService.getUserByEmail.mockImplementation(() => {
      return Promise.resolve(userMock);
    });
    await VehiclesService.searchVehicles.mockImplementation(() => {
      return Promise.resolve(vehiclesMock);
    });
    await VehiclesService.getCurrentLocation.mockImplementation(() => {
      return Promise.resolve(locationMock);
    });
  });

  it("snapshot renders", () => {

    const component = renderer.create(
      <ClientTracking />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should get user vehicles after search', async () => {
    const wrapper = shallow(
      <ClientTracking />
    )

    expect(wrapper.find('FormSelect').props().disabled).toBe(true)
    expect(wrapper.find('option')).toHaveLength(1)

    const mockedEvent = {
      target: {
        "search-input": {
          value: "test@test.com"
        }
      },
      preventDefault: () => {},
      persist: () => {}
    }
    wrapper.find('CardHeader').children().children().at(0).children().props().searchFunction(mockedEvent) // NavbarSearch
    
    await flushPromises();

    expect(wrapper.find('FormSelect').props().disabled).toBe(false)
    expect(wrapper.find('option')).toHaveLength(2)
  })

  it('should get vehicle locations', async () => {
    const wrapper = shallow(
      <ClientTracking />
    )

    const mockedEvent1 = {
      target: {
        "search-input": {
          value: "test@test.com"
        }
      },
      preventDefault: () => { },
      persist: () => { }
    }
    wrapper.find('CardHeader').children().children().at(0).children().props().searchFunction(mockedEvent1) // NavbarSearch

    await flushPromises();

    const mockedEvent2 = {
      target: {
        value: 1
      },
      preventDefault: () => { },
      persist: () => { }
    }
    wrapper.find('FormSelect').props().onChange(mockedEvent2)

    await flushPromises();

    expect(wrapper.find('CardBody').children().props().markers).toHaveLength(1);
    expect(wrapper.find('CardBody').children().props().markers[0].id).toBe(locationMock.id);
    expect(wrapper.find('CardBody').children().props().markers[0].position.lat).toBe(locationMock.lat);
    expect(wrapper.find('CardBody').children().props().markers[0].position.lng).toBe(locationMock.lng);
  })
});