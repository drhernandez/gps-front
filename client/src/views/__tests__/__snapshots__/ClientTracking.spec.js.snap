// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientTracking component snapshot renders 1`] = `
<div
  className="main-content-container px-4 container-fluid"
>
  <div
    className="page-header py-4 no-gutters row"
  >
    <div
      className="text-sm-left mb-3 col-sm-12 text-center text-md-left mb-sm-0 col-12 col-sm-4"
    >
      <span
        className="text-uppercase page-subtitle"
      >
        Clientes
      </span>
      <h3
        className="page-title"
      >
        Tracking
      </h3>
    </div>
  </div>
  <div
    className="py-4 no-gutters row"
  >
    <div
      className="mx-auto w-100 mb-4 card card-small"
    >
      <div
        className="mt-3 card-header"
      >
        <div
          className="row"
        >
          <div
            className="col"
          >
            <form
              className="main-navbar__search w-100 d-none d-md-flex d-lg-flex"
              onSubmit={[Function]}
            >
              <div
                className="input-group input-group-seamless"
              >
                <div
                  className="input-group-prepend"
                >
                  <span
                    className="input-group-text"
                  >
                    <i
                      className="material-icons"
                    >
                      search
                    </i>
                  </span>
                </div>
                <input
                  className="navbar-search search_box form-control"
                  disabled={false}
                  id="search-input"
                  placeholder="Buscar clientes por email..."
                />
              </div>
            </form>
          </div>
          <div
            className="col"
          >
            <form
              className="add-new-post"
            >
              <select
                className="form-control custom-select"
                defaultValue="default"
                disabled={true}
                id="feInputState"
                onChange={[Function]}
              >
                <option
                  disabled={true}
                  value="default"
                >
                  Elija un veh√≠culo...
                </option>
              </select>
            </form>
          </div>
        </div>
      </div>
      <div
        className="card-body"
      >
        <div
          style={
            Object {
              "height": "100%",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
